CLDR-11049 First version of the new lightweight data access API for ICU data conversion.

hagbard opened the pull request:
Note that this CL includes basic tests, but coverage for these will be improved later (with the aim of getting close to 100% coverage, including error handling and exceptions). The new code soon to be in the ICU project will also have a suite of end-to-end tests in the actual ICU data conversion logic.

- [x] Issue filed: https://unicode-org.atlassian.net/browse/CLDR-11049
- [x] Updated PR title and link in previous line to include Issue number

hagbard commented:
I don't know how to tell Jira that the '2' commits in this branch are really fine because the first one is the other pull request. I don't want to squash, since the other changes are separate.

hagbard commented:
I've resynced and this is now ready to review.

hagbard commented:
I've updated this with a lot of fixes around how attributes are handled. It now understands the ldmlICU "special" DTD and enforces attribute access according to whether it's optional or required. This caught a bunch of potential issues in the prototype ICU code and has removed some other internal complexities/hacks.

macchiati commented on tools/java/org/unicode/cldr/api/XmlDataSource.java:
```diff
@@ -0,0 +1,329 @@
+    // See: https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_Java
+    private final Object cacheLock = new Object();
+
+    // TODO: Handle draft status properly (it's a METADATA attribute).
```
In the canonical format, the draft status is always only on leaf nodes. We allowed for them at higher levels just to enable us to bulk import a whole subtree with a particular draft status. But once canonicalized, that doesn't occur. And shouldn't occur in any release.

hagbard replied:
Ahh, turns out this isn't true:
common/transforms/Azerbaijani-Latin-BGN.xml
where the <transform> element contains the draft status rather than the <tRule> child element. I'll leave the code as-is for now since it matches the current behaviour and we can revisit once the expected semantics have been sorted out.

srl295 replied:
^ is it something like the leaf 'distinguishing' element?

hagbard replied:
Maybe? I have no idea though, so I'll wait until someone can clarify. So far I've never had to care about "distinguishing" vs "non-distinguishing" elements, so anything done in that regard would be new territory. Can you point at any documentation on the semantic meaning of "distinguishing" vs "non-distinguishing" elements??

srl295 commented:
> I don't know how to tell Jira that the '2' commits in this branch are really fine because the first one is > the other pull request. I don't want to squash, since the other changes are separate.

It's OK and not that important. it gets squashed anyway. I will review when i get a chance.

srl295 commented on tools/java/org/unicode/cldr/api/AttributeKey.java:
```diff
@@ -0,0 +1,260 @@
+ * }</pre>
+ *
+ */
+// Note: Using Guava's @AutoValue library would remove all this boiler-plate.
```
or, make it a followon issue

hagbard replied:
> If you merge my other PR we can use java 9 modules!

What's the benefit to that right now (I'm not against it, just curious)?

> I think this is a good candidate for moving to the API and deleting references to CLDRFile.GEN_VERSION

Yeah, good point. I wonder who else uses the version string other than conversion code.

srl295 replied:
^ may be a good candidate for a followon issue.

hagbard replied:
Yes, and easy to do. The difficulty is that it introduces annotation meta-programming into the ICU project and will likely change how things have to be set up (to run code-gen during the build).

hagbard replied:
https://unicode-org.atlassian.net/projects/CLDR/issues/CLDR-13161

srl295 commented on tools/java/org/unicode/cldr/api/CldrDataSupplier.java:
```diff
@@ -0,0 +1,241 @@
+public abstract class CldrDataSupplier {
+    /** Returns the current CLDR version string (e.g. {@code "36"}). */
+    public static String getCldrVersionString() {
+        return CLDRFile.GEN_VERSION;
```
i assume this is here for API locality, a comment to that effect might be good so that it's not refactored out in the future.

hagbard replied:
Yes, there's a bunch of little things like this just so the api package is all that's ever needed.
I'm hoping that we could have a check for not including CLDR APIs in the ICU code (since it's otherwise so trivial to accidentally import things from outside). One way would be to compile against an "empty" JAR that only contains stubs for the API.

hagbard replied:
https://unicode-org.atlassian.net/browse/CLDR-13162

srl295 commented on tools/java/org/unicode/cldr/api/CldrDataType.java:
```diff
@@ -0,0 +1,204 @@
+     * Locale based LDML data consisting of internationalization information and translations on a
+     * per locale basis. LDML data for one locale may be inherited from other locales.
+     */
+    LDML(DtdType.ldml, true, true, DtdType.ldmlICU);
```
why is ICU needed here?

hagbard replied:
This was added recently since I have to support the "specials" directory, and I have to know things like "which elements are leaf elements" or "which attributes are valid". I'd rather not add it, but it allowed me to remove a whole pile of hard coded hacks if I just had access to the DTD.

The simple fact is that CLDR uses semantics above and beyond just XML, and things like CldrPath need that additional knowledge to work cleanly. I'd happily remove this if we can figure out a way to not need it, but for now it's definitely the best way to keep the API behaviour clean.

srl295 replied:
From an API perspective, we shouldn't expect the API to enforce semantics that aren't representable in the DTD for specials.  So if someone wants to use the ShoeSize.dtd extension that you've never seen before, any semantics beyond what the DTD can enforce shouldn't be enforced by this API. 

Can't the DTD be read from the actual XML?

hagbard replied:
That's sadly impossible at the moment.

The DTD can be read, but the DTD extensions are CLDR specific and the code that parses any special DTD extensions is hidden away in DtdData, which is in turn hooked to the DtdType enum, which cannot be arbitrarily extended.

It could all be done, but would require significant rewriting of underlying classes (rewrite DtdData, remove the DtdType enum in favour of some extensible ID class, hook DTD parsing into the core XML parser, probably rewrite chunks of CLDRFile etc.).

However this is all completely out-of-scope for my work, so I don't intend to consider trying to attempt it. If you feel strongly that it should be done, convince Mark to schedule another followup project for it.

srl295 replied:
Could we just perform no validation on specials…?  what do you think, @macchiati ? If this is really an API, it shouldn't be just a shim to enable ICU.

hagbard replied:
That's what my code used to do, but it doesn't work when you try and make "specials" work.

I need to be able to identify distinguishing attributes on _all_ path elements, since the CldrPath instance _only_ keeps those, and must be unique per visitation.

If you assume unknown attributes are "value" attributes, you get the same "special" path multiple times (because what should be a distinguishing attribute isn't present) but if you assume unknown attributes are distinguishing, you allow for duplicates to be treated differently (which messes up inheritance).

The semantics buried at the heart of CLDR are more than can be expressed in "normal" XML, and (without a lot of rewriting) you need the core libraries to understand all the DTDs that might be encountered. It's not a code problem so much as a fundamental data semantics problem.

hagbard replied:
Also when you say ""If this is really an API, it shouldn't be just a shim to enable ICU"", that's not true, since the entire remit of this work _is_ to make a shim that mostly just supports ICU.

There's nothing stopping people reworking all this once it's in and once people have decided what they want the semantics for these extensions to be, but right now, my stated goal is _only_ to get ICU and a small number of Google internal special cases to work.

srl295 commented on tools/java/org/unicode/cldr/api/CldrDataSupplier.java:
```diff
@@ -0,0 +1,241 @@
+        LinkedHashMultimap<CldrDataType, Path> multimap = LinkedHashMultimap.create();
+        for (CldrDataType type : CldrDataType.values()) {
+            type.getSourceDirectories()
+                .map(cldrRootDir.resolve("common")::resolve)
```
`common` should not be codified here. Perhaps there should be a setter for the array of subdirs… perhaps by default `['common']` but maximally `['common','seed','exemplars','keyboards']`

yes, I know ldml2icu doesn't work this way. but it's broken. c.f. `SimpleFactory.make(File sourceDirectory[], …)` where an array of input directories is used.

So, this function shouldn't be static, with no other config data coming in but the path.

hagbard replied:
This is an implementation detail and works fine for now, so I'd propose raising a bug (especially since I just don't have the context you do on this sort of thing). I'd be happy to extend the API to better understand some of these use cases, but for the NewIcu2LdmlConverter, this seems to be fine for now (I might realise it isn't as I convert the last few classes though, and if so, I'll fix it).

srl295 commented on tools/java/org/unicode/cldr/api/CldrDraftStatus.java:
```diff
@@ -0,0 +1,42 @@
+ */
+public enum CldrDraftStatus {
+    /** Include all values during visitation, performing no filtering. */
+    UNCONFIRMED(DraftStatus.unconfirmed),
```
API locality again. Should we re-export `DraftStatus`? I'd prefer refactoring/renaming `DraftStatus` to move it here under api. This is exactly the kind of bifurcation I don't want to see with this effort.

hagbard replied:
We could refactor the existing code to use this, though I'd say we should wait until the dust has settled for that. So far the duplication is only a few enums and a static method or two and it doesn't duplicate any business logic.

hagbard commented on tools/cldr-unittest/src/org/unicode/cldr/api/CldrFileDataSourceTest.java:
```diff
@@ -0,0 +1,62 @@
+
+        List<CldrPath> arbitraryKeyList = new ArrayList<>(arbitraryOrderMap.keySet());
+        List<CldrPath> dtdKeyList = new ArrayList<>(dtdOrderMap.keySet());
+        assertNotEquals("dtd order differs", arbitraryKeyList, dtdKeyList);
```
True enough.

hagbard commented on tools/java/org/unicode/cldr/api/CldrDataType.java:
```diff
@@ -0,0 +1,204 @@
+    }
+
+    Attribute getAttribute(String elementName, String attributeName) {
+        Attribute attr = DtdData.getInstance(mainType).getAttribute(elementName, attributeName);
```
The "getInstance()" method already caches, I checked. I think the right thing (if anything would be changed) is to document that getInstance() doesn't need caching by callers.

srl295 commented:
If you merge my other PR we can use java 9 modules!

I think this is a good candidate for moving to the API and deleting
references to CLDRFile.GEN_VERSION

srl295 commented:
The problem is that common only will then become default expected behavior.
I’d prefer an explicit setting such as common_only.

hagbard commented:
> The problem is that common only will then become default expected behavior. I’d prefer an explicit setting such as common_only.

Oh I completely agree, but right now the scope of my job is to migrate the NewIcu2LdmlConverter, and nothing more. I'm intended to do a pass over this whole area later because the conversion code I've migrated is at least an order of magnitude slower right now, but this is almost completely down to the fact that I'm repeatedly reading 10x more files than I need to. So having more control of the files/directories being scanned is important. Tweaking the API to reflect this will be needed so we should discuss it again then. Or you can propose an API change (since you have context as to what the use cases are).

srl295 commented:
Ldml2icu is opt in, With an explicit list of locales included in icu.

So actually, better default behavior may be to include seed by default
since any desired locales can be chosen.

macchiati commented:
Agreed. We're also planning to move all from seed to common.

{phone}

On Sat, Jul 13, 2019, 07:46 Steven R. Loomis <notifications@github.com>
wrote:

> Ldml2icu is opt in, With an explicit list of locales included in icu.
>
> So actually, better default behavior may be to include seed by default
> since any desired locales can be chosen.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/unicode-org/cldr/pull/73?email_source=notifications&email_token=ACJLEMACWX2B4DJKO3CKAOTP7F23FA5CNFSM4H4VFHZKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZ3LDNQ#issuecomment-511095222>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACJLEMGRNSXICFS5XPCJG6TP7F23FANCNFSM4H4VFHZA>
> .
>

hagbard commented:
Right, I think I've addressed all the comments so far. Assuming tests pass etc. I'm happy for this to be merged now, but please give the diffs/comments another look over (I don't have authority to merge myself).

srl295 commented:
@hagbard merged, thanks!
